ignite/contrib/handlers/tensorboardX.py

```diff
@@ -0,0 +1,185 @@
+                                           engine.state.iteration)
+                else:
+                    raise ValueError('engine.state.{key} should be of int or float.'.format(key=key))
+            else:
+                raise ValueError('engine.state does not have attribute %s.' % key)
+
+    def _log_evaluator(self, engine, evaluator, mode='training'):
+        if not isinstance(evaluator, Engine) and evaluator.state.metrics is not None:
+            raise ValueError('evaluator must be an instance of ignite.Engine and have metrics attached to it.')
+        else:
+            for key, value in evaluator.state.metrics.items():
+                self.writer.add_scalar(''.join([mode, '/', key]), value, engine.state.epoch)
+
+    def _log_histograms(self, engine, model, add_name=False):
+        if engine.state.epoch % self.histogram_freq == 0:
+            for name, param in model.named_parameters():
+                if add_name:
+                    name = ''.join([model.__class__.__name__, '/', name])
+
+                if param.requires_grad:
+                    self.writer.add_histogram(name,
+                                              param.cpu().data.numpy().flatten(),
+                                              engine.state.epoch)
+                    if self.write_grads:
+                        self.writer.add_histogram(name + '_grad',
+                                                  param.grad.cpu().data.numpy().flatten(),
+                                                  engine.state.epoch)
+
+    def _base_epoch_functions(self, engine):
+        # Use engine.state.metrics
+        if self.use_metrics:
+            self._log_metrics(engine)
+
+        # Train Evaluator
+        if self.train_evaluator is not None:
+            self._log_evaluator(engine=engine,
+                                evaluator=self.train_evaluator,
+                                mode='training')
+        # Validation Evaluator
+        if self.validation_evaluator is not None:
+            self._log_evaluator(engine=engine,
+                                evaluator=self.validation_evaluator,
+                                mode='validation')
+
+        # Histogram of Weights and Gradients; in the case of multiple models, users can create custom functions
+        if self.model is not None and self.histogram_freq > 0:
+            self._log_histograms(engine,
+                                 self.model,
+                                 add_name=False)
+
+    def _base_iteration_functions(self, engine):
+        # Use user created state attributes such as 'reward' for RL
+        if self.state_keys is not None:
+            self._log_custom_state(engine)
+
+        # Use engine.state.output for example mnist.py example
+        if self.use_output:
+            self._log_engine_output(engine)
+
+    def _on_complete(self, engine):
+        self.writer.close()
+
+    def _difference(self):
```

Comment:
@anmolsjoshi could you please detail what does this method ? Name of the method is not clear

Reply:
@vfdev-5 are you referring to `_difference`? it can be named better, it basically finds user defined functions that need to added to Tensorboard Summary if they were to use this handler as a base class. These functions could be creating images, audio, video, etc.

I showed that use here in [dcgan_with_tensorboardX.py](https://github.com/anmolsjoshi/ignite/blob/tensorboard/examples/gan/dcgan_with_tensorboardX.py#L145)